### PR TITLE
[SPARK-15177] [SparkR] [ML] SparkR 2.0 QA: New R APIs and API docs for mllib.R

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

SparkR 2.0 QA: New R APIs and API docs for mllib.R. Main changes including:
- Make `spark.glm` and `spark.naiveBayes` API more consistent with Spark naming convention. Because most Spark MLlib algorithms do not override the base R functions, we can make the argument names consistent with Spark MLlib rather than base R. Meanwhile, make the default value to be consistent with MLlib.

From:

```
setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
          function(data, formula, family = gaussian, epsilon = 1e-06, maxit = 25) {})
setMethod("spark.naiveBayes", signature(data = "SparkDataFrame", formula = "formula"),
          function(data, formula, laplace = 0, ...) {})
```

To:

```
setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
          function(data, formula, family = gaussian, tol = 1e-06, maxIter = 25) {})
setMethod("spark.naiveBayes", signature(data = "SparkDataFrame", formula = "formula"),
          function(data, formula, smoothing = 1.0, ...) {}) 
```
- Refactor the structure of mllib.R, and it will make developers and contributors understand code more clearly.

```
spark.glm, glm, summary, predict
spark.survreg, summary, predict
spark.naiveBayes, summary, predict
spark.kmeans, summary, fitted, predit
write.ml, read.ml
```
- Remove duplicated test code in test_mllib.R. Because `glm` internally will call `spark.glm` to train model, we should not duplicate all `spark.glm` tests twice and just run the basic test case to check the API working well. 
- Fix and update docs.
- Formatting code.
## How was this patch tested?

Existing unit tests.

### CODE REVIEW

File: R/pkg/R/mllib.R [hunk 1]
```diff
@@ -108,28 +107,32 @@ setMethod(
...
 #' data(iris)
 #' df <- createDataFrame(sqlContext, iris)
-#' model <- glm(Sepal_Length ~ Sepal_Width, df, family="gaussian")
+#' model <- glm(Sepal_Length ~ Sepal_Width, df, family = "gaussian")
 #' summary(model)
 #' }
 setMethod("glm", signature(formula = "formula", family = "ANY", data = "SparkDataFrame"),
           function(formula, family = gaussian, data, epsilon = 1e-06, maxit = 25) {
             spark.glm(data, formula, family, epsilon, maxit)
           })
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Because ```glm``` is R-compliant function, so I left the argument names consistent with native R.

File: R/pkg/R/mllib.R [hunk 2]
```diff
@@ -195,9 +198,9 @@ print.summary.GeneralizedLinearRegressionModel <- function(x, ...) {
     "Number of Fisher Scoring iterations: ", x$iter, "\n", sep = "")
   cat("\n")
   invisible(x)
-  }
+}
 
-#' Make predictions from a generalized linear model
+#' Makes predictions from a fitted MLlib model
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: This line appeared in the ```Description``` section of each ```rdname```, so ```predict``` for all models will share this description and it should be general.

File: R/pkg/R/mllib.R [hunk 3]
```diff
@@ -377,32 +476,9 @@ setMethod("predict", signature(object = "KMeansModel"),
             return(dataFrame(callJMethod(object@jobj, "transform", newData@sdf)))
           })
 
-#' Fit a Bernoulli naive Bayes model
+#' Saves a fitted MLlib model to the input path
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Ditto with ```predict```, it shared by all ```write.ml``` and should be general.

**Comment by yanboliang** [comment]
cc @mengxr

File: R/pkg/R/mllib.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/mllib.R]: We need have an agreement on the style of the title for a method. That is, whether or not to add "s" suffix to a verb. Seems R's tradition is not. for example, the title for "lapply":
    Apply a Function over a List or Vector

Comment (yanboliang) [inline R/pkg/R/mllib.R]: Sounds good, updated. Thanks!

File: R/pkg/R/mllib.R [hunk 4]
```diff
@@ -269,9 +349,29 @@ setMethod("summary", signature(object = "NaiveBayesModel"),
...
+#'}
+setMethod("predict", signature(object = "NaiveBayesModel"),
+          function(object, newData) {
+            return(dataFrame(callJMethod(object@jobj, "transform", newData@sdf)))
+          })
+
 #' Fit a k-means model
 #'
-#' Fit a k-means model, similarly to R's kmeans().
+#' Fits a k-means model, similarly to R's kmeans().
```
Comment (felixcheung) [inline R/pkg/R/mllib.R]: seems to be changing a few times here - should this be `Fits a ...` or `Fit a ...`?

Comment (mengxr) [inline R/pkg/R/mllib.R]: `Fits`. See https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html for examples.

Comment (felixcheung) [inline R/pkg/R/mllib.R]: I'm referring to line
```
#' Fit a k-means model
#'
#' Fit a k-means model, similarly to R's kmeans().
```
ie. the first line and the third line.

For example, it shows up for `glm` like this http://spark.apache.org/docs/latest/api/R/glm.html
image

Which I'd think would be rather odd if they are not consistent.

**Comment by vectorijk** [comment]
Suggested by this [comment](https://github.com/apache/spark/pull/13394#issuecomment-222560187), I was wondering if we also need to update the docs for k-means and naive bayes in [http://people.apache.org/~pwendell/spark-nightly/spark-master-docs/latest/sparkr.html](url). Maybe we can include that change in this PR.

**Comment by yanboliang** [comment]
@vectorijk There is a separate PR focus on updating machine learning section of SparkR users guide. FYI #13285. Thanks.

**Comment by shivaram** [comment]
@mengxr @yanboliang Is this PR still active ? Just checking if this is something we should track for the 2.0 release

**Comment by mengxr** [comment]
It would be nice to get this in. @yanboliang is traveling. I can help send a PR based on this one.

**Comment by mengxr** [comment]
@yanboliang We are going to split the work into multiple PRs (SPARK-16090). Do you mind closing this PR for now? Thanks!

**Pull request closed without merging**